cluster/images/etcd/migrate/data_dir.go

```diff
@@ -0,0 +1,163 @@
+		}
+	}
+	dir, err := os.Open(path)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open data directory %s: %s", path, err)
+	}
+
+	versionFile := &VersionFile{
+		path: filepath.Join(path, versionFilename),
+	}
+	return &DataDirectory{path, dir, versionFile}, nil
+}
+
+// Close releases the underlying File used by the DataDirectory.
+func (d *DataDirectory) Close() error {
+	return d.dir.Close()
+}
+
+// Initialize set the version.txt to the target version if the data
+// directory is empty. If the data directory is non-empty, no
+// version.txt file will be written since the actual version of etcd
+// used to create the data is unknown.
+func (d *DataDirectory) Initialize(target *EtcdVersionPair) error {
+	isEmpty, err := d.IsEmpty()
+	if err != nil {
+		return err
+	}
+	if isEmpty {
+		glog.Infof("data directory '%s' is empty, writing target version '%s' to version.txt", d.path, target)
+		err = d.versionFile.Write(target)
+		if err != nil {
+			return fmt.Errorf("failed to write version.txt to '%s': %s", d.path, err)
+		}
+		return nil
+	}
+	return nil
+}
+
+// Backup creates a backup copy of data directory.
+func (d *DataDirectory) Backup() error {
+	backupDir := fmt.Sprintf("%s.bak", d.path)
+	err := os.RemoveAll(backupDir)
+	if err != nil {
+		return err
+	}
+	err = os.MkdirAll(backupDir, 0755)
+	if err != nil {
+		return err
+	}
+	err = exec.Command("cp", "-r", d.path, backupDir).Run()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// IsEmpty returns true if the data directory is entirely empty.
+func (d *DataDirectory) IsEmpty() (bool, error) {
+	_, err := d.dir.Readdirnames(1)
+	if err == io.EOF {
+		return true, nil
+	}
+	return false, err
+}
+
+// String returns the data directory path.
+func (d *DataDirectory) String() string {
+	return d.path
+}
+
+// VersionFile provides utilities for reading and writing version.txt files
+// to etcd "data-dir" for tracking the etcd server and storage verions
+// of the data in the directory.
+type VersionFile struct {
+	path string
+}
+
+// Exists returns true if a version.txt file exists on the filesystem.
+func (v *VersionFile) Exists() (bool, error) {
+	return exists(v.path)
+}
+
+// Read parses the version.txt file and returns it's contents.
+func (v *VersionFile) Read() (*EtcdVersionPair, error) {
+	data, err := ioutil.ReadFile(v.path)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read version file %s", v.path)
+	}
+	txt := strings.TrimSpace(string(data))
+	vp, err := ParseEtcdVersionPair(txt)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse etcd '<version>/<storage-version>' string from version.txt file contents '%s': %s", txt, err)
+	}
+	return vp, nil
+}
+
+// Write creates or overwrites the contents of the version.txt file with the given EtcdVersionPair.
+func (v *VersionFile) Write(vp *EtcdVersionPair) error {
+	data := []byte(fmt.Sprintf("%s/%s", vp.version, vp.storageVersion))
+	return ioutil.WriteFile(v.path, data, 0644)
```

Comment:
Similar to above, bash default behavior is `0666` minus umask.  Not sure (a) if `WriteFile` respects umask nor (b) what our default umask has been.

Reply:
@cheftako @mml, Thanks for bringing this up. I learned something new today! 
I did some reading and according to the docs golang applies the umask to whatever permissions are provided here, so the correct thing for me to do is pass in `0666` here and `0777` for the `Mkdir*` calls and let umask be applied as appropriate.